src/OpenCV/VideoIO/Types.hsc

```diff
@@ -0,0 +1,300 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module OpenCV.VideoIO.Types
+    ( -- VideoCodecs
+      FourCC (..)
+    , VideoCaptureProperties (..)
+    , marshalCaptureProperties
+    , VideoCaptureAPI (..)
+    , marshalVideoCaptureAPI
+    ) where
+
+
+import "base" Data.Int ( Int32 )
+import "base" Data.Word ( Word32 )
+import "base" Data.String ( IsString (..) )
+import "base" Data.List( unfoldr )
+
+import qualified "inline-c" Language.C.Inline as C
+import qualified "inline-c-cpp" Language.C.Inline.Cpp as C
+import "this" OpenCV.Internal.C.Inline ( openCvCtx )
+import "this" OpenCV.Internal.C.Types
+import "this" OpenCV.Internal.Core.Types.Mat
+import "this" OpenCV.Internal.Exception
+import "this" OpenCV.TypeLevel
+
+--------------------------------------------------------------------------------
+
+C.context openCvCtx
+
+C.include "opencv2/core.hpp"
+C.include "opencv2/videoio.hpp"
+C.using "namespace cv"
+
+#include <bindings.dsl.h>
+#include "opencv2/core.hpp"
+#include "opencv2/videoio.hpp"
+
+#include "namespace.hpp"
+
+--------------------------------------------------------------------------------
+-- FourCC
+--------------------------------------------------------------------------------
+
+newtype FourCC = FourCC { unFourCC :: Int32 }
+   deriving (Enum, Num)
+
+instance IsString FourCC where
+  fromString "ask" = -1
+  fromString fcc = fromIntegral (foldr (\x acc -> acc * 256 + (toEnum . fromEnum) x) 0 fcc :: Word32)
+
+instance Show FourCC where
+  show (FourCC (-1)) = "ask"
+  show (FourCC fcc) = take 4 $ unfoldr worker (fromIntegral fcc :: Word32)
+    where
+      worker 0 = Nothing
+      worker b = Just (toEnum . fromEnum $ b `mod` 256 , b `div` 256)
+
+--------------------------------------------------------------------------------
+-- VideoCaptureProperties
+-- more at https://github.com/opencv/opencv/blob/master/modules/videoio/include/opencv2/videoio.hpp
+--------------------------------------------------------------------------------
+
+data VideoCaptureProperties     -- ^
```

Comment:
This comment is a Haddock error. We only want to put the `-- ^` on the constructors.

```
src/OpenCV/VideoIO/Types.hsc:65:33: error:
    parse error on input ‘-- ^’
```